src/DynamoRevit/DynamoRevitApp.cs

```diff
@@ -336,6 +344,42 @@ public static Assembly ResolveAssembly(object sender, ResolveEventArgs args)
             }
         }
 
+        /// <summary>
+        /// Handler when an assembly is loaded into Revit's appdomain - we need to make sure
+        /// that another addin has not loadead another version of a .dll that we require.
+        /// If this happens Dynamo will most likely crash. We should alert the user they
+        /// have an incompatible addin installed.
+        /// TODO(potentially use an additional appdomain to work around this).
+        /// </summary>
+        /// <param name="assembly"></param>
+        /// <param name="assemblyNamesToIgnore"></param>
+        /// <returns></returns>
+        public static bool AppDomainHasMismatchedReferences(Assembly assembly, String[] assemblyNamesToIgnore)
+        {
+            //get all assemblies that are currently loaded into the appdomain.
+            var loadedAssemblies = AppDomain.CurrentDomain.GetAssemblies().Select(x => x.GetName()).ToList();
+            // ignore some assemblies(revit assemblies) that we know work and have changed their version number format or do not align
+            // with semantic versioning.
+            loadedAssemblies.RemoveAll(x => assemblyNamesToIgnore.Contains(x.Name));
+            //build dict- ignore those with duplicate names.
+            var loadedAssemblyDict = loadedAssemblies.GroupBy(assm => assm.Name).ToDictionary(g => g.Key, g => g.First());
+
+            foreach (var currentAssembly in assembly.GetReferencedAssemblies().Concat(new AssemblyName[] { assembly.GetName() }))
+            {
```

Comment:
Based on our discussion, I am still curious about the N square complexity here. I'll try to install the profiling tool tomorrow. If not happening, we could also put timers around and do a couple run. Thanks for pulling this together.